core/subscription/subscriber.hpp

```diff
@@ -0,0 +1,89 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_SUBSCRIBER_HPP
+#define KAGOME_SUBSCRIBER_HPP
+
+#include <functional>
+#include <memory>
+#include <mutex>
+
+#include "subscription_engine.hpp"
+
+namespace kagome::subscription {
+
+  template <typename Key, typename Type, typename... Arguments>
+  class Subscriber final : public std::enable_shared_from_this<
+                               Subscriber<Key, Type, Arguments...>> {
+   public:
+    using KeyType = Key;
+    using ValueType = Type;
+    using HashType = size_t;
+
+    using SubscriptionEngineType =
+        SubscriptionEngine<KeyType, ValueType, Arguments...>;
+    using SubscriptionEnginePtr = std::shared_ptr<SubscriptionEngineType>;
+
+   private:
+    using SubscriptionsContainer =
+        std::unordered_map<KeyType,
+                           typename SubscriptionEngineType::IteratorType>;
+
+    SubscriptionEnginePtr engine_;
+    ValueType object_;
+
+    std::mutex subscriptions_cs_;
+    SubscriptionsContainer subscriptions_;
+
+    std::function<void(ValueType&, const KeyType &, const Arguments &...)> on_notify_callback_;
+
+   public:
+    template <typename... Args>
+    explicit Subscriber(SubscriptionEnginePtr &ptr, Args &&... args)
+        : engine_(ptr), object_(std::forward<Args>(args)...) {}
+
+    ~Subscriber() {
+      /// Unsubscribe all
+      for (auto &[key, it] : subscriptions_) engine_->unsubscribe(key, it);
+    }
+
+    Subscriber(const Subscriber &) = delete;
+    Subscriber &operator=(const Subscriber &) = delete;
+
+    Subscriber(Subscriber &&) = default;
+    Subscriber &operator=(Subscriber &&) = default;
+
+    void set_callback(std::function<void(ValueType&, const KeyType &, const Arguments &...)> &&f) {
```

Comment:
Naming

Reply:
?

Reply:
According to our code style methods are to be named in camel case.